### PR TITLE
AP_Periph: fixed CUAV_GPS LED control

### PR DESCRIPTION
**Pull request by tridge**
This fixes several problems with neopixel control in CUAV_GPS:
 - API error in AP_Periph
 - incorrect setup of timer for neopixel DMA
 - lack of event for LED send

This also fixes a bug that could cause a watchdog with neopixel enabled, as seen on f405-MatekGPS. The double call to init() meant that the dshot timer was cleared when active, which resulted in a null ptr dereference

@andyp1per this should preserve your timing for DShot, but restores the old timing for other DMA handling. I suspect this may have impacted some blheli pass-thru too

### CODE REVIEW

File: libraries/AP_HAL_ChibiOS/RCOutput.cpp [hunk 1]
```diff
@@ -762,9 +763,19 @@ bool RCOutput::setup_group_DMA(pwm_group &group, uint32_t bitrate, uint32_t bit_
         group.pwm_started = false;
     }
 
-    // original prescaler calculation from betaflight. bi-dir dshot is incredibly sensitive to the bitrate
     const uint32_t target_frequency = bitrate * bit_width;
-    uint32_t prescaler = uint32_t(lrintf((float) group.pwm_drv->clock / (bitrate * bit_width) + 0.01f) - 1);
+    uint32_t prescaler = group.pwm_drv->clock / target_frequency;
```
Comment (andyp1per) [inline libraries/AP_HAL_ChibiOS/RCOutput.cpp:767]: Might be clearer to just declare prescaler here and then move the assignment inside the else clause

File: libraries/AP_HAL_ChibiOS/RCOutput.cpp [hunk 2]
```diff
@@ -762,9 +763,19 @@ bool RCOutput::setup_group_DMA(pwm_group &group, uint32_t bitrate, uint32_t bit_
...
+    if (is_dshot) {
+        // original prescaler calculation from betaflight. bi-dir dshot is incredibly sensitive to the bitrate
+        prescaler = uint32_t(lrintf((float) group.pwm_drv->clock / (bitrate * bit_width) + 0.01f) - 1);
+    } else {
+        // adjust frequency to give an allowed value given the clock.
+        const uint32_t clock_hz = group.pwm_drv->clock;
+        while ((clock_hz / prescaler) * prescaler != clock_hz && prescaler <= 0x8000) {
```
Comment (andyp1per) [inline libraries/AP_HAL_ChibiOS/RCOutput.cpp:774]: I think this needs to be < clock_hz otherwise they may never match

File: libraries/AP_HAL_ChibiOS/RCOutput.cpp [hunk 3]
```diff
@@ -134,7 +135,7 @@ void RCOutput::rcout_thread()
     // dshot is quite sensitive to timing, it's important to output pulses as
     // regularly as possible at the correct bitrate
     while (true) {
-        chEvtWaitOne(EVT_PWM_SEND | EVT_PWM_SYNTHETIC_SEND);
+        chEvtWaitOne(EVT_PWM_SEND | EVT_PWM_SYNTHETIC_SEND | EVT_LED_SEND);
```
Comment (andyp1per) [inline libraries/AP_HAL_ChibiOS/RCOutput.cpp:138]: This will trigger dshot at irregular intervals and mess up the timing and counting. Why is this necessary? If you want to trigger an out of cycle LED update then you probably need to prevent the code below this treating it as a dshot cycle.
Note that the proper solution here (on my list) is to have a thread per timer group.

**Review by andyp1per — CHANGES_REQUESTED** [review]

File: libraries/AP_HAL_ChibiOS/RCOutput.cpp (hunk 1 above)
Comment (tridge) [reply libraries/AP_HAL_ChibiOS/RCOutput.cpp:767]: ok

File: libraries/AP_HAL_ChibiOS/RCOutput.cpp (hunk 2 above)
Comment (tridge) [reply libraries/AP_HAL_ChibiOS/RCOutput.cpp:774]: i just restored the old, well-tested, code that is known to work with LEDs and serial

File: libraries/AP_HAL_ChibiOS/RCOutput.cpp (hunk 3 above)
Comment (tridge) [reply libraries/AP_HAL_ChibiOS/RCOutput.cpp:138]: its needed as without any dshot activity the thread gets stuck waiting for an event that never happens

Comment (tridge) [reply libraries/AP_HAL_ChibiOS/RCOutput.cpp]: I've added a check for the PWM event bits for running the dshot code

**Comment by tridge** [comment]
@andyp1per I've resolved the issues, but can you test with dshot and bi-dir dshot?

**Comment by hendjoshsr71** [comment]
This also fixes a boot issue and an issue where arming would cause (reboots) with periphs that include neopixel drivers by default such as the Matek F405GPS and DLVR airspeed sensors.

I believe it will also might fix  #19803  @patrickpoirier51

**Comment by MallikarjunSE** [comment]
Hey @tridge F412 notify LEDs working. 
https://photos.app.goo.gl/BT4xvKa9ExTFt5xK7

**Comment by patrickpoirier51** [comment]
@tridge   confirmed Pixracer_periph neopixel notification works -  Thanks

**Comment by andyp1per** [comment]
> I've resolved the issues, but can you test with dshot and bi-dir dshot?

Will do, but skiing at the moment so it will be monday.

The code looks correct - I'm guessing that basic problem was that if you weren't using dshot the LEDs didn't get any cycles.

**Review by andyp1per — APPROVED** [review]

**Comment by andyp1per** [comment]
This breaks serial passthrough on H7.
